projector-client-web/src/main/kotlin/org/jetbrains/projector/client/web/speculative/Typing.kt

```diff
@@ -145,27 +145,39 @@ sealed class Typing {
         rect(
           x = editorMetrics.x,
           y = editorMetrics.y,
-          w = editorMetrics.width - 10,  // todo: -10 to prevent scrollbar be repainted
+          w = editorMetrics.width,
           h = editorMetrics.height
         )
         clip()
 
-        val imageData = getImageData(
-          sx = firstCaretLocation.x,
-          sy = firstCaretLocation.y,
-          sw = editorMetrics.width - (firstCaretLocation.x - editorMetrics.x),
-          sh = currentCarets.nominalLineHeight.toDouble() + 2 + 2
-        )
-
-        putImageData(imageData, firstCaretLocation.x + currentCarets.plainSpaceWidth, firstCaretLocation.y)
-
         val fontFace = currentCarets.fontId?.toFontFaceName() ?: "Arial"
         val fontSize = "${currentCarets.fontSize}px"
 
         font = "$fontSize $fontFace"  // todo: use a proper font style
         fillStyle = "#888"  // todo: use a proper color
 
-        val stringYPos = firstCaretLocation.y + currentCarets.fontSize  // todo: offsetting by fontSize seems to be not exact
+        val speculativeCharWidth = measureText(event.char.toString()).width
+
+        val sw = editorMetrics.width - (firstCaretLocation.x - editorMetrics.x) - currentCarets.scrollBarWidth - speculativeCharWidth
+
+        if (sw > 0) { // check we actually have some graphics to move
+          val imageData = getImageData(
+            sx = firstCaretLocation.x,
+            sy = firstCaretLocation.y,
+            sw = sw,
+            sh = currentCarets.lineHeight.toDouble()
+          )
+
+          putImageData(imageData, firstCaretLocation.x + speculativeCharWidth, firstCaretLocation.y)
+        }
+
+        val caretOffsetFromEditorEnd = editorMetrics.x + editorMetrics.width - firstCaretLocation.x
+        // check there is space to draw speculative symbol
+        if (caretOffsetFromEditorEnd < speculativeCharWidth) return@apply
```

Comment:
1. `From` -> `To` just to match the direction (if it's `From`, it would be a negative value).
2. Replace comment with an extra var.

```suggestion
        val caretToEditorEndOffset = editorMetrics.x + editorMetrics.width - firstCaretLocation.x
        val enoughSpace = speculativeCharWidth <= caretToEditorEndOffset
        if (!enoughSpace) return@apply
```

By the way, why do we need this check? Can't we just draw a part of char (setting clipping appropriately)? Well, I don't have a preference here, we can leave this check, but I wonder if there is a special reason.

Reply:
Actually I don't remember what it supposed to fix or other reasons I didn't want to draw a part of a char. I rechecked and found no problems with this check disabled so I removed it entirely